xdsl/dialects/utils/reshape_ops_utils.py

```diff
@@ -0,0 +1,54 @@
+"""
+Utilities used by reshape ops.
+See [MLIR counterpart](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h)
+for more details.
+"""
+
+from dataclasses import dataclass
+
+from typing_extensions import TypeVar
+
+from xdsl.dialects.builtin import I64, Annotated, ArrayAttr, IntegerAttr
+from xdsl.ir import Attribute
+from xdsl.irdl import AtLeast, AttrConstraint, ConstraintContext, GenericAttrConstraint
+from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.hints import isa
+
+_ReassociationAttrBase = ArrayAttr[
+    ArrayAttr[
+        Annotated[
+            IntegerAttr[I64],
+            IntegerAttr.constr(value=AtLeast(0)),
+        ]
+    ]
+]
+
+
+@dataclass(frozen=True)
+class ContiguousArrayOfIntArray(GenericAttrConstraint[_ReassociationAttrBase]):
+    """
+    Enforce an ArrayAttr of ArrayAttr[IntegerAttr] to contain contiguous integer values across all inner arrays.
+    For example: [[0, 1], [2, 3]] is valid, but [[3, 4], [0, 1]] is not.
+    An empty inner array is considered contiguous.
+    """
+
+    def verify(
+        self, attr: Attribute, constraint_context: ConstraintContext | None = None
+    ) -> None:
+        if not isa(attr, ArrayAttr[ArrayAttr[IntegerAttr]]):
```

Comment:
I would actually move the I64 thing in here, and have the _ReassociationAttrBase be a plain array of array of integer attr